backend/de.metas.monitoring/src/main/java/de/metas/monitoring/adapter/MicrometerPerformanceMonitoringService.java

```diff
@@ -42,63 +42,88 @@
 public class MicrometerPerformanceMonitoringService implements PerformanceMonitoringService
 {
 	private final MeterRegistry meterRegistry;
-	private final Optional<APMPerformanceMonitoringService> apmPerformanceMonitoringService;
+	private static final ThreadLocal<PerformanceMonitoringData> perfMonDataTL = ThreadLocal.withInitial(PerformanceMonitoringData::new);
 
 	public MicrometerPerformanceMonitoringService(
-			@NonNull final Optional<APMPerformanceMonitoringService> apmPerformanceMonitoringService,
 			@NonNull final MeterRegistry meterRegistry)
 	{
-		this.apmPerformanceMonitoringService = apmPerformanceMonitoringService;
 		this.meterRegistry = meterRegistry;
 	}
 
 	@Override
-	public <V> V monitorTransaction(
+	public <V> V monitor(
 			@NonNull final Callable<V> callable,
-			@NonNull final TransactionMetadata metadata)
+			@NonNull final PerformanceMonitoringService.Metadata metadata)
 	{
-		final ArrayList<Tag> tags = createTagsFromLabels(metadata.getLabels());
-		mkTagIfNotNull("Name", metadata.getName()).ifPresent(tags::add);
 
-		final Callable<V> callableToUse;
-		if (apmPerformanceMonitoringService.isPresent())
+		final PerformanceMonitoringData perfMonData = perfMonDataTL.get();
+		final ArrayList<Tag> tags = new ArrayList<>();
+
+		mkTagIfNotNull("name", metadata.getClassName()).ifPresent(tags::add);
+		mkTagIfNotNull("action", metadata.getFunctionName()).ifPresent(tags::add);
+
+		if(perfMonData.getDepth() == 0
```

Comment:
> perfMonData.getDepth() == 0

introduce a method in perfMonData class for that logic.
something like isFirstCall or isFirstDepth ... i don't like my suggestions... trying finding a better name.

Reply:
my first thought is to replace it with `!isCalledByMonitoredFunction()`